app/queries/drug_stocks_query.rb

```diff
@@ -0,0 +1,119 @@
+class DrugStocksQuery
+  def initialize(facilities:, for_end_of_month:)
+    @facilities = facilities
+    @for_end_of_month = for_end_of_month
+    # assuming that all facilities on the page have the same protocol
+    @protocol = @facilities.first.protocol
+  end
+
+  def protocol_drugs_by_category
+    @drug_categories ||= @protocol.protocol_drugs
+      .where(stock_tracked: true)
+      .order(:name, :dosage)
+      .sort_by {|protocol_drug| [protocol_drug.name, protocol_drug.dosage.to_i]}
+      .group_by(&:drug_category)
+      .sort_by {|(drug_category, _) | drug_category}
+      .to_h
+  end
+
+  def call
+    patient_counts = Patient.where(assigned_facility: @facilities.map(&:id)).group(:assigned_facility).count
+    drug_stocks = DrugStock.latest_for_facilities(@facilities, @for_end_of_month)
+
+    @facilities.each_with_object({}) do |facility, report|
+      facility_patient_count = patient_counts[facility] || 0
+      report[facility.id] = drug_stock_for_facility(facility, facility_patient_count, drug_stocks[facility.id])
+                              .merge(facility: facility)
+    end
+  end
+
+  def drug_stock_for_facility(facility, patient_count, facility_drug_stocks)
+    facility_report = {}
+    facility_report[:patient_count] = patient_count
+    return facility_report if facility_drug_stocks.nil?
+
+    drug_stocks = facility_drug_stocks.group_by { |drug_stock| drug_stock.protocol_drug.drug_category }
+    protocol_drugs_by_category.each do |(drug_category, _protocol_drugs)|
+      patient_days = patient_days(facility, drug_category, drug_stocks[drug_category], patient_count)
+      facility_report[drug_category] = patient_days.merge(drug_stocks: drug_stocks_by_drug_id(drug_stocks[drug_category]))
+    end
+    facility_report
+  end
+
+  def drug_stocks_by_rxnorm_code(drug_stocks)
+    drug_stocks.each_with_object({}) { |drug_stock, acc|
+      acc[drug_stock.protocol_drug.rxnorm_code] = drug_stock
+    }
+  end
+
+  def drug_stocks_by_drug_id(drug_stocks)
+    drug_stocks.each_with_object({}) { |drug_stock, acc|
+      acc[drug_stock.protocol_drug.id] = drug_stock
+    }
+  end
+
+  def patient_days(facility, drug_category, drug_stocks, patient_count)
+    coefficients = patient_days_coefficients[facility.state]
+    stocks_on_hand = stocks_on_hand(coefficients, drug_category, drug_stocks)
+    new_patient_coefficient = coefficients[:drug_categories][drug_category][:new_patient_coefficient]
+    estimated_patients = patient_count * coefficients[:load_factor] * new_patient_coefficient
+    {stocks_on_hand: stocks_on_hand,
+     patient_count: patient_count,
+     load_factor: coefficients[:load_factor],
+     new_patient_coefficient: new_patient_coefficient,
+     estimated_patients: estimated_patients,
+     patient_days: (stocks_on_hand.map { |stock| stock[:stock_on_hand] }.reduce(:+) / estimated_patients).floor}
+  rescue
+    # either drug stock is nil, or drug is not in formula
```

Comment:
Can we manually report the exception to Sentry? We should know what is going wrong, especially if we're doing a blanket `rescue`.

Reply:
Yup, that was the plan here. We're planning on extracting the patient formula calculation to the model space, will pick this up along with that.

Reply:
Fixed in 3f3b63c61c00a1a0595608fb9ce543e6bb995796.